src/Processors/Transforms/Streaming/GlobalAggregatingTransform.cpp

```diff
@@ -103,14 +127,17 @@ std::pair<bool, bool> GlobalAggregatingTransform::executeOrMergeColumns(Chunk &
     if (params->emit_changelog)
     {
         assert(!params->only_merge);
-
-        auto & retracted_variants = many_data->getField<ManyRetractedDataVariants>()[current_variant];
-        auto & aggregated_variants = many_data->variants[current_variant];
-
         /// Blocking finalization during execution on current variant
         std::lock_guard lock(variants_mutex);
-        return params->aggregator.executeAndRetractOnBlock(
-            chunk.detachColumns(), 0, num_rows, *aggregated_variants, *retracted_variants, key_columns, aggregate_columns, no_more_keys);
+
+        /// Enable retract after first finalization
+        auto retract_enabled = many_data->getField<bool>();
+        if (retract_enabled) [[likely]]
+            return params->aggregator.executeAndRetractOnBlock(
+                chunk.detachColumns(), 0, num_rows, variants, key_columns, aggregate_columns, no_more_keys);
+        else
+            return params->aggregator.executeOnBlock(
+                chunk.detachColumns(), 0, num_rows, variants, key_columns, aggregate_columns, no_more_keys);
```

Comment:
as mentioned elsewhere, maybe we can remove no_more_keys optimization for streaming aggregation